src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -1,167 +1,292 @@
+    }
+  }
+
+  public void recover(TopicPartition topicPart) {
     try {
-      RecordWriter<Long, SinkRecord> writer = getWriter(topicPart, record);
-      writer.write(System.currentTimeMillis(), record);
-      if (!offsets.containsKey(topicPart)) {
-        offsets.put(topicPart, record.getOffset() - 1);
-      }
-      updateRecordCounter(topicPart);
-      if (shouldRotate(topicPart)) {
-        rotate(topicPart);
+      switch (states.get(topicPart)) {
+        case RECOVERY_STARTED:
+          pause(topicPart);
+          nextState(topicPart);
+        case RECOVERY_PARTITION_PAUSED:
+          createWAL(topicPart);
+          nextState(topicPart);
+        case WAL_CREATED:
+          applyWAL(topicPart);
+          nextState(topicPart);
+        case WAL_APPLIED:
+          truncateWAL(topicPart);
+          nextState(topicPart);
+        case WAL_TRUNCATED:
+          resetOffsets(topicPart);
+          nextState(topicPart);
+        case OFFSET_RESET:
+          resume(topicPart);
+          nextState(topicPart);
       }
-    } catch (IOException e) {
-      throw new CopycatRuntimeException(e);
+    } catch (CopycatException e) {
+      log.error("Recovery failed.");
+      setRetryBackoff(backOffMs);
     }
   }
 
-  public void close() throws IOException, CopycatException {
-    for (TopicPartition topicPart: writers.keySet()) {
-      rotate(topicPart);
+  private void execute(TopicPartition topicPart) throws IOException {
+    Queue<SinkRecord> records = buffer.get(topicPart);
+    while(!records.isEmpty()) {
+      try {
+        switch (states.get(topicPart)) {
+          case RECOVERY_STARTED:
+            pause(topicPart);
+            nextState(topicPart);
+          case RECOVERY_PARTITION_PAUSED:
+            createWAL(topicPart);
+            nextState(topicPart);
+          case WAL_CREATED:
+            applyWAL(topicPart);
+            nextState(topicPart);
+          case WAL_APPLIED:
+            truncateWAL(topicPart);
+            nextState(topicPart);
+          case WAL_TRUNCATED:
+            resetOffsets(topicPart);
+            nextState(topicPart);
+          case OFFSET_RESET:
+            resume(topicPart);
+            nextState(topicPart);
+          case WRITE_STARTED:
+            pause(topicPart);
+            nextState(topicPart);
+          case WRITE_PARTITION_PAUSED:
+            writeRecord(topicPart, records.peek());
+            records.poll();
+            if (shouldRotate(topicPart)) {
+              nextState(topicPart);
+            } else {
+              break;
+            }
+          case SHOULD_ROTATE:
+            closeTempfile(topicPart);
+            nextState(topicPart);
+          case TEMP_FILE_CLOSED:
+            appendToWAL(topicPart);
+            nextState(topicPart);
+          case WAL_APPENDED:
+            commitFile(topicPart);
+            nextState(topicPart);
+          case FILE_COMMITTED:
+            setState(topicPart, State.WRITE_PARTITION_PAUSED);
```

Comment:
Why does this go to `WRITE_PARTITION_PAUSED` if we `resume(topicPart)` immediately after setting this flag?